api/api.go

```diff
@@ -893,6 +894,16 @@ func (c *Client) query(endpoint string, out interface{}, q *QueryOptions) (*Quer
 	parseQueryMeta(resp, qm)
 	qm.RequestTime = rtt
 
+	ct, _, _ := mime.ParseMediaType(resp.Header.Get("Content-Type"))
+	if resp.StatusCode >= 400 && ct == "text/plain" {
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return qm, err
+		}
+
+		return qm, fmt.Errorf("unexpected response: %s (%d)", string(body), resp.StatusCode)
```

Comment:
Since we expect the caller to look at the error code, I think we should expose the `resp.StatusCode` as a field on an error type. Something like this:

```go
type StatusError {
    err error
    Code int
}

func (e *StatusError) Error() string {
    return fmt.Sprintf("unexpected response (%d): %v", e.Code, e.err)
}
```

That would allow callers to use https://pkg.go.dev/errors#As to get the exact error code out, without relying on string comparison (which can be error prone).

The `StatusError` could be used here as:

```go
return qm, &StatusError{Code: resp.StatusCode, err: errors.New(string(body))}
```

Reply:
Coincidentally, there was just a post on the forums asking for exactly this: https://discuss.hashicorp.com/t/golang-consul-api-pulling-out-the-statuscode-from-deeply-nested-responses/18037

Reply:
After digging a bit more, I found there are some API endpoints that handle this better already. For example, anything that uses `requireOk`: https://github.com/hashicorp/consul/blob/master/api/api.go#L1007-L1019

Maybe this `query` function should use `requireOk`, and we can change `requireOk` to use this new type?